File: src/lessons/lander/lvl3_fly_the_lander/FlyTheLanderEntity.java
package lessons.lander.lvl3_fly_the_lander;

import lessons.lander.universe.LanderEntity;
import plm.universe.Point;

public class FlyTheLanderEntity extends LanderEntity {
  @Override public void run()
  {
    initialize();
    while (isFlying()) {
      step();
      simulateStep();
    }
  }

  /* BEGIN TEMPLATE */
  /* BEGIN HIDDEN */
  double targetStart;
  double targetEnd;
  /* END HIDDEN */
  public void initialize()
  {
    /* BEGIN HIDDEN */
    Point[] ground  = getGround();
    Point lastPoint = ground[0];
    for (Point point : ground) {
      if (point != lastPoint && lastPoint.y() == point.y()) {
        targetStart = lastPoint.x();
        targetEnd   = point.x();
        return;
      }
      lastPoint = point;
    }
    /* END HIDDEN */
  }

  public void step()
  {
    /* BEGIN SOLUTION */
    if (getX() < targetStart) {
      setDesiredAngle(-30);
    } else if (getX() > targetEnd) {
      setDesiredAngle(30);
    } else {
      if (getSpeedX() > 5) {
        setDesiredAngle(25);
      } else if (getSpeedX() < -5) {
        setDesiredAngle(-25);
      } else {
        setDesiredAngle(0);
      }
    }
    setDesiredThrust(getSpeedY() < -9 ? 4 : 3);
    /* END SOLUTION */
  }
  /* END TEMPLATE */
}

File: src/lessons/lander/lvl3_fly_the_lander/FlyTheLanderEntity.py
from RemoteLander import *
def run():
    (startPos,endPos) = (0,0)
    
    # BEGIN TEMPLATE
    def initialize():
    # BEGIN SOLUTION
        nonlocal startPos
        nonlocal endPos
        lastPoint = getGround()[0]
        for point in getGround():
            if (point != lastPoint):
                if lastPoint.y == point.y:
                    return (lastPoint.x,point.x)
            lastPoint = point
        return (0, 0)
        # END SOLUTION
    # END TEMPLATE
    
    def step():
    # BEGIN HIDDEN
        if (getX() < startPos):
            setDesiredAngle(-30)
        elif (getX() > endPos):
            setDesiredAngle(30)
        elif (getSpeedX() > 5):
            setDesiredAngle(25)
        elif (getSpeedX() < -5):
            setDesiredAngle(-25)
        else:
            setDesiredAngle(0);
            
        if (getSpeedY() <-9):
            setDesiredThrust(4)
        else:
            setDesiredThrust(3)    
    # END HIDDEN
    
    
    initialize()
    while isFlying():
      step()
      simulateStep()
